### ISSUE TITLE
default format spec is not correct and compatible with the web: "WARNING: Requested formats are incompatible for merge and will be merged into mkv."

### ISSUE THREAD

**Issue by baptx** [ytdl-org/youtube-dl#29195]
## Checklist

- [x] I'm reporting a broken site support issue
- [x] I've verified that I'm running youtube-dl version **2021.05.16**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar bug reports including closed ones
- [x] I've read bugs section in FAQ

## Verbose log

```
youtube-dl -v https://www.youtube.com/watch?v=2MLVYEwXGL0
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'https://www.youtube.com/watch?v=2MLVYEwXGL0']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2021.05.16
[debug] Python version 2.7.18 (CPython) - Linux-5.4.0-73-generic-x86_64-with-Ubuntu-20.04-focal
[debug] exe versions: avconv 4.2.4, avprobe 4.2.4, ffmpeg 4.2.4, ffprobe 4.2.4, phantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[youtube] 2MLVYEwXGL0: Downloading webpage
[debug] Default format spec: bestvideo+bestaudio/best
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[debug] Invoking downloader on u'https://r1---sn-4g5e6nss.googlevideo.com/videoplayback?expire=1622648120&ei=2FC3YMnSMNWox_AP1LadkA0&ip=51.77.211.2&id=o-APSXKz4qvxK81EtI2p2MlnRGmN4mhjMoLmCLWF6LDnTm&itag=243&aitags=133%2C134%2C160%2C242%2C243%2C278&source=youtube&requiressl=yes&mh=xG&mm=31%2C29&mn=sn-4g5e6nss%2Csn-4g5ednsk&ms=au%2Crdu&mv=m&mvi=1&pl=22&initcwndbps=502500&vprv=1&mime=video%2Fwebm&ns=oEUa8bshYmeffCtshJN1YggF&gir=yes&clen=136048605&dur=5283.720&lmt=1606839162849799&mt=1622625892&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=WEB&txp=5432434&n=5WDh9u0sC2GzsKta9P&sparams=expire%2Cei%2Cip%2Cid%2Caitags%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRQIhALc67oqZqDkRqkSGjhiDfo6CUnOwyIdZHORT2i_YN6A_AiArYqLRB1s5FwZ2SuFa0nn2Pl_qxnLrtvBP4g28wWwSjQ%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIhAMxrdZpTK-yeZ_TIXD3V5Flx-vS99-z2A1_Yeu6gDMcfAiA-Snj0gFlq33IHzbxAjJSihejXgR1xFMNVhFS2S7C2RA%3D%3D'
[download] Destination: Le Bonheur au Travail   Documentaire 2015-2MLVYEwXGL0.f243.webm
[download] 100% of 129.75MiB in 01:11
[debug] Invoking downloader on u'https://r1---sn-4g5e6nss.googlevideo.com/videoplayback?expire=1622648120&ei=2FC3YMnSMNWox_AP1LadkA0&ip=51.77.211.2&id=o-APSXKz4qvxK81EtI2p2MlnRGmN4mhjMoLmCLWF6LDnTm&itag=140&source=youtube&requiressl=yes&mh=xG&mm=31%2C29&mn=sn-4g5e6nss%2Csn-4g5ednsk&ms=au%2Crdu&mv=m&mvi=1&pl=22&initcwndbps=502500&vprv=1&mime=audio%2Fmp4&ns=oEUa8bshYmeffCtshJN1YggF&gir=yes&clen=85513196&dur=5283.793&lmt=1606796218524973&mt=1622625892&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=WEB&txp=5432434&n=5WDh9u0sC2GzsKta9P&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRQIgT379yR0BaTgNqToUOuQqinMyFot2P3kl4YkaWYMW0joCIQCoP6BA_cmATKumtG3EjezkvtOa6Hr5gkHSi_79dxuBog%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIhAMxrdZpTK-yeZ_TIXD3V5Flx-vS99-z2A1_Yeu6gDMcfAiA-Snj0gFlq33IHzbxAjJSihejXgR1xFMNVhFS2S7C2RA%3D%3D'
[download] Destination: Le Bonheur au Travail   Documentaire 2015-2MLVYEwXGL0.f140.m4a
[download] 100% of 81.55MiB in 00:45
[ffmpeg] Merging formats into "Le Bonheur au Travail   Documentaire 2015-2MLVYEwXGL0.mkv"
[debug] ffmpeg command line: ffmpeg -y -loglevel 'repeat+info' -i 'file:Le Bonheur au Travail   Documentaire 2015-2MLVYEwXGL0.f243.webm' -i 'file:Le Bonheur au Travail   Documentaire 2015-2MLVYEwXGL0.f140.m4a' -c copy -map '0:v:0' -map '1:a:0' 'file:Le Bonheur au Travail   Documentaire 2015-2MLVYEwXGL0.temp.mkv'
Deleting original file Le Bonheur au Travail   Documentaire 2015-2MLVYEwXGL0.f243.webm (pass -k to keep)
Deleting original file Le Bonheur au Travail   Documentaire 2015-2MLVYEwXGL0.f140.m4a (pass -k to keep
```

## Description

The default behavior of youtube-dl is not correct and compatible with the web. For example when we want to download a video like https://www.youtube.com/watch?v=2MLVYEwXGL0, there is a message "WARNING: Requested formats are incompatible for merge and will be merged into mkv.".
This results in a webm video with an m4a audio merged into MKV. Usually webm video should be used with webm audio and mp4 video should be used with m4a audio. Web browsers don't support MKV so it will not be playable on the web. This requires additional conversion and time to make it compatible with the web, like archive.org is doing when uploading MKV files.

To fix the issue currently, we have to add parameters manually.
If we want to to download a webm file, it will be:
`youtube-dl -f 'bestvideo[ext=webm]+bestaudio[ext=webm]' https://www.youtube.com/watch?v=2MLVYEwXGL0` (in this example, it also works with fewer parameters like `youtube-dl -f 'bestvideo[ext=mp4]+bestaudio' https://www.youtube.com/watch?v=2MLVYEwXGL0`)
And if we want to to download an mp4 file, it will be:
`youtube-dl -f 'bestvideo[ext=mp4]+bestaudio[ext=m4a]' https://www.youtube.com/watch?v=2MLVYEwXGL0` (in this example, it also works with fewer parameters like `youtube-dl -f 'bestvideo+bestaudio[ext=webm]' https://www.youtube.com/watch?v=2MLVYEwXGL0`)

I suggest that the default behavior is to download mp4 files since it is compatible with most devices. Or at least avoid the mkv warning by choosing a video/audio combination that is compatible either with mp4 or webm.

Update: with the command `youtube-dl -F https://www.youtube.com/watch?v=2MLVYEwXGL0`, it looks like there is a file with better quality (ID 18):
```
[youtube] 2MLVYEwXGL0: Downloading webpage
[info] Available formats for 2MLVYEwXGL0:
format code  extension  resolution note
249          webm       audio only tiny   49k , webm_dash container, opus @ 49k (48000Hz), 30.92MiB
250          webm       audio only tiny   62k , webm_dash container, opus @ 62k (48000Hz), 39.14MiB
251          webm       audio only tiny  122k , webm_dash container, opus @122k (48000Hz), 76.85MiB
140          m4a        audio only tiny  129k , m4a_dash container, mp4a.40.2@129k (44100Hz), 81.55MiB
160          mp4        256x144    144p   51k , mp4_dash container, avc1.4d400c@  51k, 25fps, video only, 32.73MiB
278          webm       256x144    144p   79k , webm_dash container, vp9@  79k, 25fps, video only, 49.76MiB
133          mp4        426x240    240p  101k , mp4_dash container, avc1.4d4015@ 101k, 25fps, video only, 64.18MiB
242          webm       426x240    240p  112k , webm_dash container, vp9@ 112k, 25fps, video only, 70.58MiB
134          mp4        640x360    360p  186k , mp4_dash container, avc1.4d401e@ 186k, 25fps, video only, 117.48MiB
243          webm       640x360    360p  205k , webm_dash container, vp9@ 205k, 25fps, video only, 129.75MiB
18           mp4        640x360    360p  447k , avc1.42001E, 25fps, mp4a.40.2 (44100Hz), 281.95MiB (best)
```

The correct way to get it is `youtube-dl -f best https://www.youtube.com/watch?v=2MLVYEwXGL0` so I think this should be the default behavior when available.

Related: https://github.com/ytdl-org/youtube-dl/issues/26810 (but more a question than issue)

I also noticed an issue on https://github.com/ytdl-org/youtube-dl#format-selection (found on https://github.com/ytdl-org/youtube-dl/issues/26892#issuecomment-709303960):
```
# Download best mp4 format available or any other best if no mp4 available
$ youtube-dl -f 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best'
```
For the video mentioned previously, it will not get the best quality (ID 18 with 447k).

**Comment by dirkf** [comment]
In the [linked issue](https://github.com/ytdl-org/youtube-dl/issues/26892#issuecomment-709628742):
`--merge-output-format mp4`

**Comment by baptx** [comment]
@dirkf yes, this is another workaround but I am not sure if it is the proper one to have the best compatibility because it is merging webm video with m4a audio in mp4 container (instead of webm video + webm audio in webm or mp4 video + m4a audio in mp4). The best solution would be a default behavior suggested previously instead of the warning and having to use additional parameters manually.

**Comment by baptx** [comment]
@dstftw I see you closed the issue, was the default behavior changed? A solution to this issue without having to use additional parameters would be the best user experience.

**Comment by dirkf** [comment]
If you want some supported behaviour as default, you just have to set the parameters for that behaviour in one of the documented configuration files. For instance:
```
--format 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/bestvideo+bestaudio/best'
--merge-output-format mp4
```
The default behaviour is clearly documented, as well as how to override it.

**Comment by baptx** [comment]
@dirkf I meant it would be nice if it is the default configuration for everybody (without having to change configuration files) so we can share a command and the other user will get the same quality. But even `--format 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/bestvideo+bestaudio/best'` does not get the best quality for some videos, for example for the link I shared, it will be a file size of 117.48MiB but to get the file with the size of 281.95MiB, we need to do `youtube-dl -f best https://www.youtube.com/watch?v=2MLVYEwXGL0`.

**Comment by dirkf** [comment]
In the end, you can't guarantee to get the format you want unless you list the formats first (`-F`) and then specify your selected format (or combination of formats) explicitly (`-f`).

**Comment by baptx** [comment]
@dirkf by the way, the option `--merge-output-format mp4` did not work when the video is mp4 and audio is webm, for example with the command `youtube-dl --merge-output-format mp4 https://www.youtube.com/watch?v=VKjxbQbhoZo`:
```
[youtube] VKjxbQbhoZo: Downloading webpage
[youtube] VKjxbQbhoZo: Downloading MPD manifest
[download] Destination: Longest Surf Days at Alaïa Bay-VKjxbQbhoZo.f137.mp4
[download] 100% of 77.74MiB in 00:42
[download] Destination: Longest Surf Days at Alaïa Bay-VKjxbQbhoZo.f251.webm
[download] 100% of 2.65MiB in 00:01
[ffmpeg] Merging formats into "Longest Surf Days at Alaïa Bay-VKjxbQbhoZo.mp4"
ERROR:   Stream #1:0 -> #0:1 (copy)
```

**Comment by dirkf** [comment]
WFM: downloaded, merged (ffmpeg 4.3-2), plays with video and sound in `mpv` based on ffmpeg 4.2.2
```
$ youtube-dl --merge-output-format mp4 'https://www.youtube.com/watch?v=VKjxbQbhoZo'
[youtube] VKjxbQbhoZo: Downloading webpage
[youtube] VKjxbQbhoZo: Downloading MPD manifest
[download] Destination: Longest Surf Days at Alaïa Bay-VKjxbQbhoZo.f137.mp4
[download] 100% of 77.74MiB in 01:07
[download] Destination: Longest Surf Days at Alaïa Bay-VKjxbQbhoZo.f251.webm
[download] 100% of 2.65MiB in 00:02
[ffmpeg] Merging formats into "Longest Surf Days at Alaïa Bay-VKjxbQbhoZo.mp4"
Deleting original file Longest Surf Days at Alaïa Bay-VKjxbQbhoZo.f137.mp4 (pass -k to keep)
Deleting original file Longest Surf Days at Alaïa Bay-VKjxbQbhoZo.f251.webm (pass -k to keep)
$
```
